File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;


/**
0. ) size가 0이면 false
1. ( 일 때만 스택 push
2. ) 일 때만 스택 pop
3. 최종적으로 stack size가 0이면 true
*/
class Solution {
    boolean solution(String s) {
        Stack<Character> st = new Stack();
        
        for (char ch: s.toCharArray()) {
            if (st.isEmpty() && ch == ')') return false;
            else if (ch == '(') st.push(ch);
            else if (ch == ')') st.pop();
            
        }
        
        if (st.isEmpty()) return true;
        else return false;
        
        
    }
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.cpp
#include <bits/stdc++.h>

using namespace std;


bool solution(string s) {
    stack<int> st;
    
    for (int i = 0; i < (int) s.length(); i++) {
        char ch = s[i];
        if (!st.empty() && ch == ')' && st.top() == '(') {
            st.pop();
        } else {
            st.push(ch);
        }
    }
    
    if (st.empty()) return true;

    return false;
}